libnetdata/gorilla/gorilla.cc

```diff
@@ -0,0 +1,618 @@
+    //     - find its leading-zero count
+
+    Word xor_value = bc->prev_number ^ number;
+    // FIXME: Use SFINAE
+    Word xor_lzc = (bit_size<Word>() == 32) ? __builtin_clz(xor_value) : __builtin_clzll(xor_value);
+    Word is_xor_lzc_same = (xor_lzc == bc->prev_xor_lzc) ? 1 : 0;
+
+    if (is_xor_lzc_same) {
+        // xor-lzc is same
+        if (bit_stream_write(bs, static_cast<Word>(1), 1) == false) {
+            goto RET_FALSE;
+        }
+    } else {
+        // xor-lzc is different
+        if (bit_stream_write(bs, static_cast<Word>(0), 1) == false) {
+            goto RET_FALSE;
+        }
+        
+        if (bit_stream_write(bs, xor_lzc, (bit_size<Word>() == 32) ? 5 : 6) == false) {
+            goto RET_FALSE;
+        }
+    }
+
+    // write the bits of the XOR value without the LZC prefix
+    if (bit_stream_write(bs, xor_value, bit_size<Word>() - xor_lzc) == false) {
+        goto RET_FALSE; 
+    }
+
+    bc->prev_number = number;
+    bc->prev_xor_lzc = xor_lzc;
+    return true;
+
+RET_FALSE:
+    bc->bs.position = position;
+    return false;
+}
+
+// only valid for writers
+template<typename Word>
+static bool bit_code_flush(bit_code_t<Word> *bc) {
+    bit_stream_t<Word> *bs = &bc->bs;
+
+    Word num_entries_written = bc->entries;
+    Word num_bits_written = bs->position;
+
+    // we want to write these at the beginning
+    bs->position = 0;
+
+    if (!bit_stream_write(bs, num_entries_written, bit_size<Word>())) {
+        return false;
+    }
+
+    if (!bit_stream_write(bs, num_bits_written, bit_size<Word>())) {
+        return false;
+    }
+
+    bs->position = num_bits_written;
+    return true;
+}
+
+// only valid for readers
+template<typename Word>
+static bool bit_code_info(bit_code_t<Word> *bc, Word *num_entries_written,
+                                                Word *num_bits_written) {
+    bit_stream_t<Word> *bs = &bc->bs;
+
+    assert(bs->position == 2 * bit_size<Word>());
+    if (bs->capacity < (2 * bit_size<Word>())) {
+        return false;
+    }
+
+    if (num_entries_written) {
+        *num_entries_written = bs->buffer[0];
+    }
+    if (num_bits_written) {
+        *num_bits_written = bs->buffer[1];
+    }
+
+    return true;
+}
+
+template<typename Word>
+static size_t gorilla_encode(Word *dst, Word dst_len, const Word *src, Word src_len) {
+    bit_code_t<Word> bcw;
+
+    bit_code_init(&bcw, dst, dst_len);
+
+    for (size_t i = 0; i != src_len; i++) {
+        if (!bit_code_write(&bcw, src[i]))
+            return 0;
+    }
+
+    if (!bit_code_flush(&bcw))
+        return 0;
+
+    return src_len;
+}
+
+template<typename Word>
+static size_t gorilla_decode(Word *dst, Word dst_len, const Word *src, Word src_len) {
+    bit_code_t<Word> bcr;
+
+    bit_code_init(&bcr, (Word *) src, src_len);
+
+    Word num_entries;
+    if (!bit_code_info(&bcr, &num_entries, (Word *) NULL)) {
+        return 0;
+    }
+    if (num_entries > dst_len) {
+        return false;
```

Comment:
`false` is zero, but considering that this is the unique return that is not explicitly returning a size, I suggest to return 0 here too.